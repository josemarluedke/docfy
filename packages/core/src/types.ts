/* eslint-disable @typescript-eslint/no-explicit-any */

import { Node as MarkdownAST } from 'unist';
import { Processor, Plugin as UnifiedPlugin, Settings as UnifiedSettings } from 'unified';
import { VFile } from 'vfile';

export interface Heading {
  title: string;
  id: string;
  depth: number;
  headings?: Heading[];
}

export interface PageMetadata {
  url: string;
  relativeUrl: undefined | string;
  relativePath: string;
  editUrl: string;
  title: string;
  headings: Heading[];
  frontmatter: Record<string, unknown>;
  pluginData: Record<string, unknown>;
  parentLabel: undefined | string;
}

export interface PageContent {
  meta: PageMetadata;
  sourceConfig: SourceConfig;
  source: string;
  vFile: VFile;
  ast: MarkdownAST;
  markdown: string;
  rendered: string;
  demos?: PageContent[];
  pluginData: Record<string, unknown>;
}

interface ContextOptions
  extends Omit<Options, 'plugins' | 'remarkPlugins' | 'rehypePlugins' | 'tocMaxDepth'> {
  tocMaxDepth: number;
}

export interface StaticAssetDefinition {
  fromPath: string;
  toPath: string;
}

export interface Context {
  remark: Processor;
  rehype: Processor;
  pages: PageContent[];
  staticAssets: StaticAssetDefinition[];
  options: ContextOptions;
}

export interface NestedPageMetadata {
  name: string;
  label: string;
  pages: PageMetadata[];
  children: NestedPageMetadata[];
}

export interface DocfyResult {
  content: PageContent[];
  staticAssets: StaticAssetDefinition[];
  nestedPageMetadata: NestedPageMetadata;
}

export interface SourceConfig {
  /**
   * The absolute path to where the files are located.
   */
  root: string;

  /**
   * Match files using the patterns the shell uses, like stars and stuff. It
   * uses Glob package.
   */
  pattern: string;

  /**
   * Pattern to ignore.
   */
  ignore?: string[];

  /**
   * Option to change how urls are generated.
   *
   * There are two ways URLs can ge generated.
   *
   * 1. "auto": Uses the folder structure to inform how the urls are generated.
   *    For example, if you have the folowing files:
   *      - install.md
   *      - components/
   *        - button.md
   *        - card.md
   *    The urls would look like this, (assuming urlPrefix is set to "docs").
   *      - docs/install
   *      - docs/components/buttons
   *      - docs/components/card
   *
   * 2. "manual": It uses frontmatter information to inform "subcategory" and
   *    "category" of the file, ignoring the original file location.
   *    The url schema is as follows:
   *    {category}/{subcategory}/{file-name}
   *
   *    If no category or subcategory is specified, all files will be at the root
   *    level.
   *
   *    This options is perfect for documenting monorepo projects while keeping
   *    documentation next to implementation.
   */
  urlSchema?: 'auto' | 'manual';

  /**
   * Option to prefix the urls.
   *
   * For example:
   * "docs" will generate urls like "/docs/something"
   * "blog" will generate urls like "/blog/something"
   */
  urlPrefix?: string;

  /**
   * Option to suffix the urls.
   *
   * For example:
   * ".html" will generate urls like "/something.html"
   */
  urlSuffix?: string;

  /**
   * Overwrite repository config for this source.
   */
  repository?: RepositoryConfig;
}

interface RepositoryConfig {
  /**
   *
   * The url to the Git Repository
   * Example: https://github.com/josemarluedke/docfy
   */
  url: string;

  /**
   * Branch used to edit your markdown when clicking on "Edit this page" button.
   * @default "master"
   */
  editBranch?: string;
}

export interface PluginOptions {
  [key: string]: unknown;
}

export type PluginHandler<T = PluginOptions | undefined | null> = (
  ctx: Context,
  options: T
) => void;

export interface Plugin<T = PluginOptions | undefined | null> {
  runBefore?: PluginHandler<T>;
  runWithMdast?: PluginHandler<T>;
  runWithHast?: PluginHandler<T>;
  runAfter?: PluginHandler<T>;
}

export interface PluginWithOptions<T = PluginOptions> extends Plugin<T> {
  __options?: T;
}

export type PluginWithOptionsFunction<T = PluginOptions> = (options?: T) => PluginWithOptions<T>;

export type PluginList = (Plugin | PluginWithOptions<any> | PluginWithOptionsFunction<any>)[];

export interface Options {
  /**
   * A list of Docfy plugins.
   */
  plugins?: PluginList;

  /**
   * Additional remark plugins
   *
   * Example:
   *
   * ```js
   * const hbs = require('remark-hbs');
   * const autolinkHeadings = require('remark-autolink-headings');
   *
   * const remarkPlugins = [
   *   [
   *     autolinkHeadings,
   *     {
   *       behavior: 'wrap'
   *     }
   *   ],
   *   hbs
   * ];
   * ```
   */
  remarkPlugins?: ([UnifiedPlugin, UnifiedSettings] | UnifiedPlugin)[];

  /**
   * Additional rehype plugins
   *
   * Example:
   *
   * ```js
   * const rehypePrism = require('@mapbox/rehype-prism');
   *
   * const rehypePlugins = [rehypePrism];
   * ```
   */
  rehypePlugins?: ([UnifiedPlugin, UnifiedSettings] | UnifiedPlugin)[];

  /**
   * The max depth of headings
   * @default 6
   */
  tocMaxDepth?: number;

  /**
   * The repository config
   */
  repository?: RepositoryConfig;

  /**
   * Labels to be used while generating nestedPageMetadata
   */
  labels?: Record<string, string>;

  /**
   * The static asset path to be used in the their url.
   * Eg. staticAssetsPath: "/assets/docfy" and a static asset name as
   * "github-icon.png", the URL to be used for the image would be
   * "/assets/docfy/github-icon.png".
   *
   * @default "/assets/docfy"
   */
  staticAssetsPath?: string;
}

interface DocfyConfigSourceConfig extends Omit<SourceConfig, 'root'> {
  root?: string;
}

export interface DocfyConfig extends Options {
  sources: DocfyConfigSourceConfig[];
}
