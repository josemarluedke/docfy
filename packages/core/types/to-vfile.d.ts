declare module 'to-vfile' {
  import { VFile } from 'vfile';

  class ToVFile {
    static readSync(filepath: string): VFile;
  }

  export default ToVFile;
}
