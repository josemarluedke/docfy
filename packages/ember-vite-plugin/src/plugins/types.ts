import { Node } from 'unist';

interface Literal {
  value: string;
}

export interface CodeNode extends Node, Literal {
  type: 'code';
  lang?: string;
  meta?: string;
}

export interface DemoComponent {
  name: DemoComponentName;
  chunks: DemoComponentChunk[];
  description?: {
    title?: string;
    ast: Node;
    editUrl?: string;
  };
}

export interface DemoComponentName {
  dashCase: string;
  pascalCase: string;
}

export interface DemoComponentChunk {
  type: string;
  code: string;
  ext: string;
  snippet: Node;
}
