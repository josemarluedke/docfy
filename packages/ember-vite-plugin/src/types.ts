import type { Node } from 'unist';

interface Literal {
  value: string;
}
export interface CodeNode extends Node, Literal {
  type: 'code';
  lang?: string;
  meta?: string;
}
export interface ImportStatement {
  type: 'component' | 'helper' | 'modifier' | 'service' | 'other';
  name: string;
  path: string;
  isDefault: boolean;
  namedImports?: string[];
}

export interface InlineComponent {
  name: string;
  type: 'template-only' | 'class-based' | 'const-template';
  template: string;
  script?: string;
  imports?: ImportStatement[];
}

export interface GJSMetadata {
  imports: ImportStatement[];
  inlineComponents: InlineComponent[];
  additionalContent: string;
  templateContent: string; // The main HTML/HBS content
}

// Legacy demo component format from ember package
export interface LegacyDemoComponent {
  name: {
    pascalCase: string;
    dashCase: string;
  };
  chunks: Array<{
    ext: string;
    code: string;
    type: string;
  }>;
}

// New demo component format matching original ember implementation
export interface DemoComponentName {
  dashCase: string;
  pascalCase: string;
}

export interface DemoComponentChunk {
  type: string;
  code: string;
  ext: string;
  snippet: any; // AST node reference
}

export interface DemoComponent {
  name: DemoComponentName;
  chunks: DemoComponentChunk[];
  description?: {
    title?: string;
    ast: any; // AST node reference
    editUrl?: string;
  };
}

export interface PreviewTemplateComponent {
  name: string;
  template: string;
  ext: string; // File extension (hbs, gjs, gts, etc.)
  originalNode?: any; // Reference to AST node for replacement
}
