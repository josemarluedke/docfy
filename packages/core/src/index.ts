import glob from 'fast-glob';
import path from 'path';
import trough, { Through } from 'trough';
import toVfile from 'to-vfile';
import {
  PageContent,
  Context,
  Options,
  SourceConfig,
  PageMetadata,
  DocfyResult,
  Plugin,
  PluginList,
  PluginWithOptions,
  PluginWithOptionsFunction,
} from './types';
import { DEFAULT_IGNORE, generateAutoUrl, generateManualUrl, inferTitle } from './-private/utils';
import { createRehype, createRemark } from './-private/remark';
import {
  combineDemos,
  renderMarkdown,
  replaceInternalLinks,
  staticAssets,
  toc,
  uniquefyUrls,
  removeUnnecessaryIndex,
} from './plugins';
import { getRepoEditUrl } from './-private/repo-info';
import { transformToNestedPageMetadata } from './-private/nested-page-metadata';
import debugFactory from 'debug';
const debug = debugFactory('@docfy/core');

class Docfy {
  private pipeline: Through<Context>;
  private context: Context;
  private plugins: PluginList;

  constructor(options: Options = {}) {
    const { remarkPlugins, rehypePlugins, ...rest } = options;

    this.context = {
      remark: createRemark(remarkPlugins),
      rehype: createRehype(rehypePlugins),
      pages: [],
      staticAssets: [],
      options: {
        ...rest,
        tocMaxDepth: rest.tocMaxDepth || 6,
      },
    };

    const plugins: PluginList = [
      combineDemos,
      removeUnnecessaryIndex,
      uniquefyUrls,
      replaceInternalLinks,
      staticAssets,
    ];

    if (Array.isArray(options.plugins)) {
      plugins.push(...options.plugins);
    }

    // Make sure TOC and renderMarkdown plugins are the last ones
    plugins.push(toc, renderMarkdown);
    this.plugins = plugins;

    this.pipeline = trough<Context>()
      .use<SourceConfig[]>(this.initializePipeline.bind(this))
      .use(this.createPluginPipelineFor('runBefore'))
      .use(this.createPluginPipelineFor('runWithMdast'))
      .use(this.transformerMdastToHast)
      .use(this.createPluginPipelineFor('runWithHast'))
      .use(this.createPluginPipelineFor('runAfter'));
  }

  public run(sources: SourceConfig[]): Promise<DocfyResult> {
    return new Promise((resolve, reject) => {
      this.pipeline.run(sources, (err: unknown, ctx: Context): void => {
        if (err) {
          reject(err);
        } else {
          resolve({
            content: ctx.pages,
            staticAssets: ctx.staticAssets,
            nestedPageMetadata: transformToNestedPageMetadata(
              ctx.pages.map(p => p.meta),
              ctx.options.labels
            ),
          });
        }
      });
    });
  }

  private transformerMdastToHast(ctx: Context): void {
    ctx.pages.forEach(page => {
      const hast = ctx.rehype.runSync(page.ast, page.vFile);
      page.ast = hast;

      page.demos?.forEach(demo => {
        const hast = ctx.rehype.runSync(demo.ast, demo.vFile);
        demo.ast = hast;
      });
    });
  }

  private createPluginPipelineFor(
    funcType: 'runBefore' | 'runAfter' | 'runWithMdast' | 'runWithHast'
  ): (ctx: Context) => void {
    function isPluginWithOptionsFunction(
      plugin: Plugin | PluginWithOptions | PluginWithOptionsFunction
    ): plugin is PluginWithOptionsFunction {
      return typeof plugin === 'function' && '__isOptionsFunction' in plugin;
    }

    function isPluginWithOptions(
      plugin: Plugin | PluginWithOptions | PluginWithOptionsFunction
    ): plugin is PluginWithOptions {
      return '__options' in plugin;
    }

    const plugins = this.plugins;

    return function (ctx: Context): void {
      plugins.forEach(plugin => {
        let options: unknown;

        if (isPluginWithOptionsFunction(plugin)) {
          plugin = plugin();
        }

        if (isPluginWithOptions(plugin)) {
          options = plugin.__options;
        }

        const fn = plugin[funcType];
        if (fn) {
          fn(ctx, options as never);
        }
      });
    };
  }

  private initializePipeline(sources: SourceConfig[]): Context {
    const ctx = this.context;

    // Clear previous results to ensure fresh content on each run
    ctx.pages = [];
    ctx.staticAssets = [];

    sources.forEach(item => {
      let repoEditUrl: string | null;

      if (item.repository || ctx.options.repository) {
        repoEditUrl = getRepoEditUrl(
          item.root,
          item.repository?.url || ctx.options.repository?.url || '',
          item.repository?.editBranch || ctx.options.repository?.editBranch
        );
      }
      const files = glob.sync(item.pattern, {
        cwd: item.root,

        ignore: [...DEFAULT_IGNORE, ...(item.ignore || [])],
        absolute: true,
      });

      debug('Source Files', files);

      files.forEach(file => {
        ctx.pages.push(this.createPage(item, file, repoEditUrl));
      });
    });

    return ctx;
  }

  private createPage(
    sourceConfig: SourceConfig,
    fullPath: string,
    repoEditUrl?: string | null
  ): PageContent {
    const relativePath = fullPath.replace(path.join(sourceConfig.root, '/'), '');

    const vFile = toVfile.readSync(fullPath);
    const markdown = vFile.contents.toString();
    const parsed = this.context.remark.parse(vFile);
    const ast = this.context.remark.runSync(parsed, vFile);

    const frontmatter: Record<string, unknown> =
      (vFile.data as { frontmatter?: Record<string, unknown> }).frontmatter || {};

    let url: string;
    let title = inferTitle(ast);
    if (typeof frontmatter.title === 'string') {
      title = frontmatter.title;
    }
    // Fallback for title
    if (typeof title === 'undefined' || title === '') {
      title = path.parse(fullPath).name;
    }

    if (typeof frontmatter.url === 'string') {
      url = generateAutoUrl(frontmatter.url, sourceConfig.urlPrefix, sourceConfig.urlSuffix);
    } else if (sourceConfig.urlSchema === 'manual') {
      url = generateManualUrl(
        relativePath,
        frontmatter,
        sourceConfig.urlPrefix,
        sourceConfig.urlSuffix
      );
    } else {
      url = generateAutoUrl(relativePath, sourceConfig.urlPrefix, sourceConfig.urlSuffix);
    }

    // Add fallback order for index pages
    if (url.length > 0 && url[url.length - 1] === '/' && typeof frontmatter.order === 'undefined') {
      (frontmatter.order as undefined | number) = -1;
    }

    let editUrl = '';

    if (repoEditUrl) {
      editUrl = repoEditUrl.replace('{filepath}', relativePath);
    }

    const metadata: PageMetadata = {
      url,
      title,
      editUrl,
      relativeUrl: undefined,
      relativePath,
      headings: [],
      frontmatter: frontmatter,
      pluginData: {},
      parentLabel: undefined,
    };

    return {
      meta: metadata,
      source: relativePath,
      sourceConfig,
      vFile,
      ast,
      markdown,
      rendered: '',
      pluginData: {},
    };
  }
}

export default Docfy;
module.exports = Docfy;
