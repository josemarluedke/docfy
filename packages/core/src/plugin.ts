import { Plugin, PluginOptions, PluginWithOptions, PluginWithOptionsFunction } from './types';

function plugin(handlers: Plugin): Plugin {
  return handlers;
}

function pluginWithOptions<T = PluginOptions>(handlers: Plugin<T>): PluginWithOptionsFunction<T> {
  function optionsFunction(options?: T): PluginWithOptions<T> {
    return {
      __options: options,
      ...handlers,
    };
  }
  optionsFunction.__isOptionsFunction = true;

  return optionsFunction;
}

plugin.withOptions = pluginWithOptions;

export default plugin;
module.exports = plugin;
