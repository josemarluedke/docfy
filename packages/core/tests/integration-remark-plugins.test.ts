import Docfy from '../src';
import { DocfyResult } from '../src/types';
import path from 'path';
import autolinkHeadings from 'remark-autolink-headings';
import math from 'remark-math';
import katex from 'rehype-katex';

const root = path.resolve(__dirname, './__fixtures__/monorepo');

describe('When proving remark plugins', () => {
  describe('When plugin has no options', () => {
    let result: DocfyResult;

    beforeAll(async () => {
      const docfy = new Docfy({
        remarkPlugins: [autolinkHeadings],
      });
      result = await docfy.run([
        {
          root,
          urlPrefix: 'docs',
          urlSchema: 'manual',
          pattern: '**/*.md',
        },
      ]);
    });

    test('it should have rendered the content with the plugin with no options', () => {
      const htmls = [];
      result.content.forEach(page => {
        htmls.push([page.source, page.rendered]);
      });

      expect(htmls).toMatchSnapshot();
    });
  });

  describe('When plugin has options', () => {
    let result: DocfyResult;

    beforeAll(async () => {
      const docfy = new Docfy({
        remarkPlugins: [[autolinkHeadings, { behavior: 'append' }]],
      });
      result = await docfy.run([
        {
          root,
          urlPrefix: 'docs',
          urlSchema: 'manual',
          pattern: '**/*.md',
        },
      ]);
    });

    test('it should have rendered the content with the plugin and their options', () => {
      const htmls = [];
      result.content.forEach(page => {
        htmls.push([page.source, page.rendered]);
      });

      expect(htmls).toMatchSnapshot();
    });
  });

  describe('Using remark math plugin', () => {
    let result: DocfyResult;

    beforeAll(async () => {
      const docfy = new Docfy({
        remarkPlugins: [math],
      });
      result = await docfy.run([
        {
          root,
          urlPrefix: 'docs',
          pattern: '**/math.md',
        },
      ]);
    });

    test('it should have used the remark math plugin', () => {
      const htmls = [];
      result.content.forEach(page => {
        htmls.push([page.source, page.rendered]);
      });

      expect(htmls).toMatchSnapshot();
    });
  });
});

describe('When proving rehype plugins', () => {
  describe('When plugin has no options', () => {
    let result: DocfyResult;

    beforeAll(async () => {
      const docfy = new Docfy({
        remarkPlugins: [math],
        rehypePlugins: [katex],
      });
      result = await docfy.run([
        {
          root,
          urlPrefix: 'docs',
          pattern: '**/math.md',
        },
      ]);
    });

    test('it should have used remark and rehype plugins', () => {
      const htmls = [];
      result.content.forEach(page => {
        htmls.push([page.source, page.rendered]);
      });

      expect(htmls).toMatchSnapshot();
    });
  });
});
