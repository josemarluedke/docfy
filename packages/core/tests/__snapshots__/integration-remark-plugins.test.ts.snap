// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When proving rehype plugins When plugin has no options it should have used remark and rehype plugins 1`] = `
[
  [
    "docs/math.md",
    "<h1 id="testing-math-remark-and-rehype-plugins">Testing math remark and rehype plugins</h1>
<p><span class="math math-inline"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi></mrow><annotation encoding="application/x-tex">x</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.43056em;vertical-align:0em;"></span><span class="mord mathnormal">x</span></span></span></span></span></p>",
  ],
]
`;

exports[`When proving remark plugins Using remark math plugin it should have used the remark math plugin 1`] = `
[
  [
    "docs/math.md",
    "<h1 id="testing-math-remark-and-rehype-plugins">Testing math remark and rehype plugins</h1>
<p><span class="math math-inline">x</span></p>",
  ],
]
`;

exports[`When proving remark plugins When plugin has no options it should have rendered the content with the plugin with no options 1`] = `
[
  [
    "README.md",
    "<h1 id="the-index"><a href="#the-index" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>The index</h1>",
  ],
  [
    "docs/about.md",
    "<h1 id="about-the-project"><a href="#about-the-project" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>About the project</h1>
<h2 id="history"><a href="#history" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>History</h2>
<p>Description.</p>
<h2 id="where-it-came-from"><a href="#where-it-came-from" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>Where it came from</h2>
<p>Description.</p>
<h3 id="but-why"><a href="#but-why" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>But why?</h3>
<p>Description.</p>
<h2 id="another-topic"><a href="#another-topic" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>Another topic</h2>
<h2 id="2-level"><a href="#2-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>2 level</h2>
<h3 id="3-level"><a href="#3-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>3 level</h3>
<h3 id="3-level-2nd"><a href="#3-level-2nd" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>3 level 2nd</h3>
<h4 id="4-level"><a href="#4-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>4 level</h4>
<h5 id="5-level"><a href="#5-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>5 level</h5>
<h6 id="6-level"><a href="#6-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>6 level</h6>
<h4 id="4-level-2nd"><a href="#4-level-2nd" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>4 level 2nd</h4>",
  ],
  [
    "docs/custom-url.md",
    "<h1 id="this-a-test-for-the-ability-to-change-the-url-using-frontmatter"><a href="#this-a-test-for-the-ability-to-change-the-url-using-frontmatter" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>This a test for the ability to change the url using frontmatter</h1>
<p>The url will still contain any prefix and suffix if specified in the source
config.</p>",
  ],
  [
    "docs/how-it-works.md",
    "<h1 id="how-it-works"><a href="#how-it-works" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>How it works?</h1>
<p>This is a description.</p>
<p><a href="/docs/about">Link using reference</a></p>",
  ],
  [
    "docs/html.md",
    "<h1 id="html-should-work"><a href="#html-should-work" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>HTML should work</h1>
<div>
  test
</div>
<main>
<p><strong>bold</strong></p>
</main>",
  ],
  [
    "docs/math.md",
    "<h1 id="testing-math-remark-and-rehype-plugins"><a href="#testing-math-remark-and-rehype-plugins" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>Testing math remark and rehype plugins</h1>
<p>$x$</p>",
  ],
  [
    "docs/static-asset.md",
    "<h1 id="this-is-a-markdown-that-has-a-image"><a href="#this-is-a-markdown-that-has-a-image" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>This is a markdown that has a image</h1>
<p><img src="/assets/docfy/github-icon.png" alt="GitHub"></p>
<p><img src="/assets/docfy/github-icon.png" alt="Alt text" title="The GitHub Icon"></p>
<p>With a reference later in the document defining the Image location:</p>",
  ],
  [
    "docs/table.md",
    "<h1 id="table-example"><a href="#table-example" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>Table example</h1>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Description</th>
<th>Type</th>
<th>Default Value</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>@scope</code></td>
<td>Filter links by a scope name</td>
<td><code>string</code> | <code>undefined</code></td>
<td></td>
</tr>
</tbody>
</table>",
  ],
  [
    "packages/package1/components/button.md",
    "<h1 id="button-component"><a href="#button-component" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>Button Component</h1>",
  ],
  [
    "packages/package1/docs/intro.md",
    "<h1 id="docs-intro"><a href="#docs-intro" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>Docs Intro</h1>",
  ],
  [
    "packages/package2/docs/overview.md",
    "<h1 id="this-is-the-overview-page"><a href="#this-is-the-overview-page" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>This is the overview page</h1>
<p>This is a <a href="/docs/category1/components/button">link to another file in a different package</a>.</p>",
  ],
  [
    "packages/package2/docs/styles.md",
    "<h1 id="how-styles-work"><a href="#how-styles-work" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>How styles work</h1>
<p>Short description.</p>
<p>Please refer to <a href="/docs/category2/overview#get-started">overview</a> to get started.</p>",
  ],
  [
    "packages/package1/components/form/index.md",
    "<h1 id="form-component"><a href="#form-component" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a>Form Component</h1>",
  ],
  [
    "packages/package1/components/form/something-else.md",
    "<p>Some other file next to a index.md</p>",
  ],
]
`;

exports[`When proving remark plugins When plugin has options it should have rendered the content with the plugin and their options 1`] = `
[
  [
    "README.md",
    "<h1 id="the-index">The index<a href="#the-index" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>",
  ],
  [
    "docs/about.md",
    "<h1 id="about-the-project">About the project<a href="#about-the-project" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>
<h2 id="history">History<a href="#history" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h2>
<p>Description.</p>
<h2 id="where-it-came-from">Where it came from<a href="#where-it-came-from" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h2>
<p>Description.</p>
<h3 id="but-why">But why?<a href="#but-why" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h3>
<p>Description.</p>
<h2 id="another-topic">Another topic<a href="#another-topic" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h2>
<h2 id="2-level">2 level<a href="#2-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h2>
<h3 id="3-level">3 level<a href="#3-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h3>
<h3 id="3-level-2nd">3 level 2nd<a href="#3-level-2nd" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h3>
<h4 id="4-level">4 level<a href="#4-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h4>
<h5 id="5-level">5 level<a href="#5-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h5>
<h6 id="6-level">6 level<a href="#6-level" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h6>
<h4 id="4-level-2nd">4 level 2nd<a href="#4-level-2nd" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h4>",
  ],
  [
    "docs/custom-url.md",
    "<h1 id="this-a-test-for-the-ability-to-change-the-url-using-frontmatter">This a test for the ability to change the url using frontmatter<a href="#this-a-test-for-the-ability-to-change-the-url-using-frontmatter" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>
<p>The url will still contain any prefix and suffix if specified in the source
config.</p>",
  ],
  [
    "docs/how-it-works.md",
    "<h1 id="how-it-works">How it works?<a href="#how-it-works" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>
<p>This is a description.</p>
<p><a href="/docs/about">Link using reference</a></p>",
  ],
  [
    "docs/html.md",
    "<h1 id="html-should-work">HTML should work<a href="#html-should-work" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>
<div>
  test
</div>
<main>
<p><strong>bold</strong></p>
</main>",
  ],
  [
    "docs/math.md",
    "<h1 id="testing-math-remark-and-rehype-plugins">Testing math remark and rehype plugins<a href="#testing-math-remark-and-rehype-plugins" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>
<p>$x$</p>",
  ],
  [
    "docs/static-asset.md",
    "<h1 id="this-is-a-markdown-that-has-a-image">This is a markdown that has a image<a href="#this-is-a-markdown-that-has-a-image" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>
<p><img src="/assets/docfy/github-icon.png" alt="GitHub"></p>
<p><img src="/assets/docfy/github-icon.png" alt="Alt text" title="The GitHub Icon"></p>
<p>With a reference later in the document defining the Image location:</p>",
  ],
  [
    "docs/table.md",
    "<h1 id="table-example">Table example<a href="#table-example" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Description</th>
<th>Type</th>
<th>Default Value</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>@scope</code></td>
<td>Filter links by a scope name</td>
<td><code>string</code> | <code>undefined</code></td>
<td></td>
</tr>
</tbody>
</table>",
  ],
  [
    "packages/package1/components/button.md",
    "<h1 id="button-component">Button Component<a href="#button-component" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>",
  ],
  [
    "packages/package1/docs/intro.md",
    "<h1 id="docs-intro">Docs Intro<a href="#docs-intro" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>",
  ],
  [
    "packages/package2/docs/overview.md",
    "<h1 id="this-is-the-overview-page">This is the overview page<a href="#this-is-the-overview-page" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>
<p>This is a <a href="/docs/category1/components/button">link to another file in a different package</a>.</p>",
  ],
  [
    "packages/package2/docs/styles.md",
    "<h1 id="how-styles-work">How styles work<a href="#how-styles-work" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>
<p>Short description.</p>
<p>Please refer to <a href="/docs/category2/overview#get-started">overview</a> to get started.</p>",
  ],
  [
    "packages/package1/components/form/index.md",
    "<h1 id="form-component">Form Component<a href="#form-component" aria-hidden="true" tabindex="-1"><span class="icon icon-link"></span></a></h1>",
  ],
  [
    "packages/package1/components/form/something-else.md",
    "<p>Some other file next to a index.md</p>",
  ],
]
`;
